### ISSUE TITLE
MD023 incorrectly checks headings inside code blocks

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#231]
## Bug Description
MD023 (heading-start-left) incorrectly flags headings that appear inside code blocks (both fenced and indented).

## Impact
- ~600 false positives when running on markdownlint test suite
- Affects any documentation with code examples containing markdown headings

## Example
```markdown
Code block with heading:
```markdown
      ### This is in a code block  # Should NOT be flagged
```

Indented code block:
      # This is code, not a heading  # Should NOT be flagged
```

## Solution
Similar to MD007 fix in #228:
1. Convert MD023 from `Rule` to `AstRule` trait
2. Add code block detection using AST
3. Skip lines within code block ranges

## Test Cases
- [ ] Headings in fenced code blocks should be ignored
- [ ] Headings in indented code blocks should be ignored
- [ ] Indented headings outside code blocks should still be detected

## Related
- Parent issue: #229
- Similar fix: #228 (MD007)
- Related fix needed: #230 (MD034)
